[ML] Transforms: Fixes available fields for sort options for latest configuration

walterra opened the pull request:
## Summary

- Fixes the transform preview header to display the heading text in any case and the copy-to-clipboard button for `latest` configurations (the copy-to-clipboard option for `pivot` is displayed within the form).
- Fix to avoid listing all fields for the `sort` option for `latest` configuration and only show `date` fields
- Fixes ambiguous form field labels

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/)
* Commit: 7b22c2b2a596ab19d9e0cef1c20198f1677d4772
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/execution/node/377/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"}]}
PIPELINE-->

peteharverson commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/hooks/use_latest_function_config.ts:
```diff
@@ -56,7 +56,8 @@ function getOptions(
   const sortFieldOptions: Array<EuiComboBoxOptionOption<string>> = indexPattern.fields
-    .filter((v) => !ignoreFieldNames.has(v.name) && v.sortable)
+    // The backend API for `latest` allows all field types for sort but the UI will be limited to `date`.
+    .filter((v) => !ignoreFieldNames.has(v.name) && v.sortable && v.type === 'date')
```
Any ideas what we should do if there are no `date` fields in the source index?

![image](https://user-images.githubusercontent.com/7405507/104947723-18ccea80-59b4-11eb-8cef-ee8a2fc96492.png)

walterra replied:
An alternative solution could be to still make all fields available but group them in the dropdown by field type, with `date` fields first, then `numbers`, then `keyword`.

Something like

```
date fields

  @timestamp
  created_at

----
number fields

  responsetime

----
other

  airline
```

Screenshot of the grouped dropdown from EUI docs:

<img src="https://user-images.githubusercontent.com/230104/104949349-1a4fe000-59bf-11eb-90a9-5269a3c6b7e9.png" width="50%" />

peteharverson replied:
As discussed, let's stick to the current implementation and only show `date` type fields in the dropdown. Advanced users can switch to the API or dev tools if they really want to use a `number` field say. And add some help text below the Sort field input, along the lines of `Select the date field to be used to identify the latest document.`

lcawl replied:
That text sounds fine to me.  I wonder if we can alternatively change the "Add a sort field...." text in the screenshot to "Add a date field ...." and maybe even "...that identifies the latest documents" if space allows.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/)
* Commit: b55d78b67453fae56aa336524ed37a7a7086a5d7

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/898385c2e230e665dc9e5ca9ee231f446213ade1) | [after](https://github.com/elastic/kibana/commit/b55d78b67453fae56aa336524ed37a7a7086a5d7) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.1MB | +57.0B |
| `transform` | 1.0MB | 1.0MB | +392.0B |
| total |  |  | +449.0B |

### History
* :broken_heart: [Build #99453](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/) failed 7b22c2b2a596ab19d9e0cef1c20198f1677d4772

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/","number":"99463","commit":"b55d78b67453fae56aa336524ed37a7a7086a5d7"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/)
* Commit: 301960f79cb148c22b5dd795108d615d1db57c5e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/898385c2e230e665dc9e5ca9ee231f446213ade1) | [after](https://github.com/elastic/kibana/commit/301960f79cb148c22b5dd795108d615d1db57c5e) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.1MB | +57.0B |
| `transform` | 1.0MB | 1.0MB | +2.0KB |
| total |  |  | +2.0KB |

### History
* :green_heart: [Build #99463](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/) succeeded b55d78b67453fae56aa336524ed37a7a7086a5d7
* :broken_heart: [Build #99453](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/) failed 7b22c2b2a596ab19d9e0cef1c20198f1677d4772

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/","number":"99463","commit":"b55d78b67453fae56aa336524ed37a7a7086a5d7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/","number":"99577","commit":"301960f79cb148c22b5dd795108d615d1db57c5e"}]}
PIPELINE-->

peteharverson commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/latest_function_form.tsx:
```diff
@@ -51,25 +57,51 @@ export const LatestFunctionForm: FC<LatestFunctionFormProps> = ({ latestFunction
+                  id="xpack.transform.stepDefineForm.sortFieldOptionsEmptyError"
+                  defaultMessage="No date fields available to sort on. To override the configuration with another field type, please copy the configuration to the clipboard using the following button and continue creating the transform in Kibana Dev Console."
+                />{' '}
+                <EuiCopy beforeMessage={copyToClipboardDescription} textToCopy={copyToClipboard}>
```
Is it possible to update the `copyToClipboard` text when the user adds unique key fields? Currently the text only seems to ever have the `source` property. It would be nice if we could get the `latest` field in there too, with the `"unique_key": ` field.

walterra replied:
Fixed in 422355e.

peteharverson replied:
This works fine, apart from when I switch the transform type from `latest` to `pivot`, in which case the 'copy to clipboard' text to the right of the Pivot preview grid still contains the text for the old `latest` config.

lcawl commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/latest_function_form.tsx:
```diff
@@ -51,25 +57,51 @@ export const LatestFunctionForm: FC<LatestFunctionFormProps> = ({ latestFunction
+            <EuiComboBox
+              fullWidth
+              placeholder={i18n.translate('xpack.transform.stepDefineForm.sortPlaceholder', {
+                defaultMessage: 'Add a sort field ...',
```
```suggestion
                defaultMessage: 'Add a date field ...',
```

lcawl commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/latest_function_form.tsx:
```diff
@@ -51,25 +57,51 @@ export const LatestFunctionForm: FC<LatestFunctionFormProps> = ({ latestFunction
+              <p>
+                <FormattedMessage
+                  id="xpack.transform.stepDefineForm.sortFieldOptionsEmptyError"
+                  defaultMessage="No date fields available to sort on. To override the configuration with another field type, please copy the configuration to the clipboard using the following button and continue creating the transform in Kibana Dev Console."
```
I think this can be shortened. For example:
```suggestion
                  defaultMessage="No date fields are available to sort on. To use another field type, copy the configuration to the clipboard and continue creating the transform in the Console."
```

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/)
* Commit: 64ea3dbca943eb83efa5b4e1d604a5b935463905

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/898385c2e230e665dc9e5ca9ee231f446213ade1) | [after](https://github.com/elastic/kibana/commit/64ea3dbca943eb83efa5b4e1d604a5b935463905) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.1MB | +57.0B |
| `transform` | 1.0MB | 1.0MB | +2.1KB |
| total |  |  | +2.2KB |

### History
* :green_heart: [Build #99577](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/) succeeded 301960f79cb148c22b5dd795108d615d1db57c5e
* :green_heart: [Build #99463](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/) succeeded b55d78b67453fae56aa336524ed37a7a7086a5d7
* :broken_heart: [Build #99453](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/) failed 7b22c2b2a596ab19d9e0cef1c20198f1677d4772

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/","number":"99463","commit":"b55d78b67453fae56aa336524ed37a7a7086a5d7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/","number":"99577","commit":"301960f79cb148c22b5dd795108d615d1db57c5e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/","number":"99620","commit":"64ea3dbca943eb83efa5b4e1d604a5b935463905"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99670/)
* Commit: dccdfdf9fc04d0a11701f6f0290d46b10bf4e08a

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/898385c2e230e665dc9e5ca9ee231f446213ade1) | [after](https://github.com/elastic/kibana/commit/dccdfdf9fc04d0a11701f6f0290d46b10bf4e08a) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.1MB | +57.0B |
| `transform` | 1.0MB | 1.0MB | +2.1KB |
| total |  |  | +2.1KB |

### History
* :green_heart: [Build #99620](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/) succeeded 64ea3dbca943eb83efa5b4e1d604a5b935463905
* :green_heart: [Build #99577](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/) succeeded 301960f79cb148c22b5dd795108d615d1db57c5e
* :green_heart: [Build #99463](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/) succeeded b55d78b67453fae56aa336524ed37a7a7086a5d7
* :broken_heart: [Build #99453](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/) failed 7b22c2b2a596ab19d9e0cef1c20198f1677d4772

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/","number":"99463","commit":"b55d78b67453fae56aa336524ed37a7a7086a5d7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/","number":"99577","commit":"301960f79cb148c22b5dd795108d615d1db57c5e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/","number":"99620","commit":"64ea3dbca943eb83efa5b4e1d604a5b935463905"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99670/","number":"99670","commit":"dccdfdf9fc04d0a11701f6f0290d46b10bf4e08a"}]}
PIPELINE-->

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/latest_function_form.tsx:
```diff
@@ -7,14 +7,20 @@
 import { LatestFunctionService } from './hooks/use_latest_function_config';
 
 interface LatestFunctionFormProps {
+  copyToClipboard: any;
```
Can you please provide a type here?

walterra replied:
Good catch, rushed it, fixed in 422355e.

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/step_define_form.tsx:
```diff
@@ -377,12 +386,21 @@ export const StepDefineForm: FC<StepDefineFormProps> = React.memo((props) => {
-          <LatestFunctionForm latestFunctionService={stepDefineForm.latestFunctionConfig} />
+          <LatestFunctionForm
+            latestFunctionService={stepDefineForm.latestFunctionConfig}
+            copyToClipboard={copyToClipboardPivot}
+            copyToClipboardDescription={copyToClipboardPivotDescription}
```
Did you consider passing the "Copy to clipboard" component as a child instead of adding extra props?

walterra replied:
It's the same prop structure how we pass it on to `DataGrid` so didn't want to change that for this PR.

walterra commented:
Changed `latestConfigMapper.toAPIConfig()` to return partial configs similar to pivots so a user can copy a config which is not complete (e.g. if no date field for sort is available) to the clipboard to continue in Dev Console. That required to update `validateLatestConfig` too. I added some jest tests to cover the changes.

walterra commented:
@lcawl thanks for the feedback, updated the texts.

peteharverson reviewed (approved):
Tested latest edits and LGTM

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99980/)
* Commit: 422355ed7564ab704f74149cc43902d51e12a93e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f718e90a815d031a37ddfb2de9872a281b47a600) | [after](https://github.com/elastic/kibana/commit/422355ed7564ab704f74149cc43902d51e12a93e) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.1MB | +57.0B |
| `transform` | 1.0MB | 1.0MB | +2.3KB |
| total |  |  | +2.3KB |

### History
* :green_heart: [Build #99670](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99670/) succeeded dccdfdf9fc04d0a11701f6f0290d46b10bf4e08a
* :green_heart: [Build #99620](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/) succeeded 64ea3dbca943eb83efa5b4e1d604a5b935463905
* :green_heart: [Build #99577](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/) succeeded 301960f79cb148c22b5dd795108d615d1db57c5e
* :green_heart: [Build #99463](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/) succeeded b55d78b67453fae56aa336524ed37a7a7086a5d7
* :broken_heart: [Build #99453](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/) failed 7b22c2b2a596ab19d9e0cef1c20198f1677d4772

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99453/","number":"99453","commit":"7b22c2b2a596ab19d9e0cef1c20198f1677d4772"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99463/","number":"99463","commit":"b55d78b67453fae56aa336524ed37a7a7086a5d7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99577/","number":"99577","commit":"301960f79cb148c22b5dd795108d615d1db57c5e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99620/","number":"99620","commit":"64ea3dbca943eb83efa5b4e1d604a5b935463905"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99670/","number":"99670","commit":"dccdfdf9fc04d0a11701f6f0290d46b10bf4e08a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99980/","number":"99980","commit":"422355ed7564ab704f74149cc43902d51e12a93e"}]}
PIPELINE-->

darnautov reviewed (approved):
LGTM
